### ISSUE TITLE
Support ingesting objects that support the Arrow PyCapsule API

### ISSUE THREAD

**Issue by jonmmease** [vega/vegafusion#498]
We could support ingesting objects that implement the [Arrow PyCapsule API](https://arrow.apache.org/docs/format/CDataInterface/PyCapsuleInterface.html).

Compared to the current support for the DataFrame Interchange Protocol, accepting objects that implement the Arrow PyCapsule API wouldn't require pyarrow, and wouldn't require converting to a pyarrow Table on the Python side.

I think we could use @kylebarron's new [pyo3-arrow](https://docs.rs/pyo3-arrow/latest/pyo3_arrow/) crate for this (since it doesn't require the pyarrow dependency). In fact, I think we could drop pyarrow as a hard dependency using this approach, since pyarrow itself supports the PyCapsule API.

---

cc @MarcoGorelli based on comment in https://github.com/vega/altair/pull/3452#issuecomment-2205819070

> I've caught up with Polars devs, and they're on board with using Altair in polars.DataFrame.plot if the plots can be done directly without going via pandas

In order for VegaFusion (which powers Vega-Altair's optional [`"vegafusion"` data transformer](https://altair-viz.github.io/user_guide/large_datasets.html#vegafusion-data-transformer)) to support polars without pyarrow (so that operations like Vega-Altair's histogram binning and aggregation are performed in the Python kernel rather than in the browser), I think we'll need polars to support the PyCapsule API as discussed in https://github.com/pola-rs/polars/issues/12530.

**Comment by kylebarron** [comment]
> I think we could drop pyarrow as a hard dependency using this approach, since pyarrow itself supports the PyCapsule API.

Yes, though it does require the user to have a relatively recent version of pyarrow.

Let me know if I can help with pyo3-arrow at all! I've only published a version support `arrow` version 52, so you'll have to upgrade before you can use pyo3-arrow. I figure you only care about importing data, not exporting data?

**Comment by jonmmease** [comment]
Thanks for chiming in @kylebarron

> I've only published a version support arrow version 52

Yeah, I need to update DataFusion and Arrow soon anyway.

>  I figure you only care about importing data, not exporting data?

Thats correct

> Yes, though it does require the user to have a relatively recent version of pyarrow.

Thanks for the call out, that's a good point.

**Comment by kylebarron** [comment]
If you pointed me to where the arrow ingest happens, I could probably make a PR for this if you'd like

**Comment by jonmmease** [comment]
Thanks for the offer!

Here's is where the pyarrow tables are imported

https://github.com/vega/vegafusion/blob/007bd44188676de7259bc02a61693b3dc7586072/vegafusion-common/src/data/table.rs#L270-L286

This is invoked from the PyO3 Rust code in:

https://github.com/vega/vegafusion/blob/007bd44188676de7259bc02a61693b3dc7586072/vegafusion-python-embed/src/lib.rs#L189-L196

I'm imagining there would be a `VegaFusionDataset.from_arrow_pycapsule` or something, then another branch in `process_inline_datasets` that would check from the PyCapsule interface and use this API. Definitely happy for you to do the whole PR, but even if you only implement this `from_arrow_pycapsule` method that would be really helpful and I can do the rest of the routing later.

If this is blocked by updating arrow-rs, I can ping this thread once that's done.

**Comment by kylebarron** [comment]
> If this is blocked by updating arrow-rs, I can ping this thread once that's done

I think that's primarily a question of whether you're ok vendoring the relevant PyCapsule code (on top of arrow-rs' FFI code). It's a relatively small amount of code ([polars pr for reference](https://github.com/pola-rs/polars/pull/17693/files#diff-e3d47a60859272b56e0db84b98a73422ab6679cf58d3be3cc7394e0ac937dcd8)), and then you don't have to add a dependency on pyo3-arrow if you don't want.
